### ISSUE TITLE
Make it flexible to add more losses and choose loses via command line

### ISSUE THREAD

**Issue by mathmanu** [ClementPinard/FlowNetPytorch#28]
I like the way models and datasets are arranged - makes this a great framework to extend for further work.

However, losses are not arranged that way. I wish if you would create a 'losses' folder for the losses and allow a list of losses (which will be added together) to be loaded by specifying it from command line. 

As you might have read, unsupervised optical is reaching accuracy close to supervised optical flow estimation. "UnFlow: Unsupervised Learning of Optical Flow with a Bidirectional Census Loss"
https://arxiv.org/abs/1711.07837
This is an amazing work that deserves lot of attention.

If we make losses in this code base flexible, may be it will be easy to added other learning tasks such as UnFlow easily into this code base.

**Comment by mathmanu** [comment]
Example - add the following to multiscaleloss.pyafter moving to losses folder:

__all__ = [
    'l2_norm_loss', 'l1_norm_loss', 'l2_norm_error'
]

def l2_norm_loss(input_var, network_output, target_flow, weights=None, sparse=False):
    return multiscaleEPE(network_output, target_flow, weights, sparse, power=2)


def l1_norm_loss(input_var, network_output, target_flow, weights=None, sparse=False):
    return multiscaleEPE(network_output, target_flow, weights, sparse, power=1)


def l2_norm_error(input_var, network_output, target, weights=None, sparse=False):
    b, _, h, w = target.size()
    upsampled_output = nn.functional.upsample(network_output, size=(h,w), mode='bilinear')
    return EPE(upsampled_output, target, sparse, mean=True, power=2)


Also support options such as --loss-functions, --error-functions in main.py so that the appropriate loss and error functions can be selected.

**Comment by mathmanu** [comment]
Once this is done, it can called from main.py as follows (of course, you know this, as you have done this in the case of models and datasets - but writing down just to be clear):


    loss = 0
    for lidx, loss_fn in enumerate(args.loss_functions):
        loss += args.div_flow * loss_functions.__dict__[loss_fn](input_var, output, target_var, \
                     weights=args.multiscale_weights, sparse=args.sparse)

    flow2_EPE = 0
    for eidx, error_fn in enumerate(args.error_functions):
        flow2_EPE += args.div_flow * loss_functions.__dict__[error_fn](input_var, output[0], target_var, \
                     weights=None, sparse=args.sparse)

**Comment by mathmanu** [comment]
And hypothetically, to lean UnFlow (unsupervised), I would specify the losses and errors as:

--loss-functions photo_loss census_loss flow_consistency_loss --error-functions l2_norm_error

**Comment by ClementPinard** [comment]
This could be a nice basis for this paper, but I feel like it would be completely different workflow. My goal here is to give a code that can be modified easily.
And it seems to me that adding all these loss functions will make the code much more heavy and less accessible for customization.

How do you deal with the absence of ground truth ? How do you deal with occultation maps ? Where do you generate them ? In the training function or in the loss ?

The code for a training algorithm such as UnFlow to my mind is much better within a fork of FlowNet than a feature of it, because it will need heavy modifications, that will make the code harder to learn from (readability is very important !)

**Comment by mathmanu** [comment]
Okay. I was just looking for an UnFlow code in PyTorch and couldn't find it. I thought this code base could be extended - but I agree with you, its not that simple. I hope some one will port it to PyTorch.
